torch/nn/quantized/modules/linear.py

```diff
@@ -83,13 +84,43 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
         super(LinearPackedParams, self)._load_from_state_dict(state_dict, prefix, local_metadata, False,
                                                               missing_keys, unexpected_keys, error_msgs)
 
-    @torch.jit.export
     def __getstate__(self):
         qweight, bias = self._weight_bias()
-        return qweight, bias, self.training, self.dtype
+        # Make a copy and save attributes from torch.nn.Module
+        state = self.__dict__.copy()
+        return qweight, bias, self.training, self.dtype, state
+
+    # Duplicate state init from torch.nn.Module.__init__, avoid calling `super` which not supported well with torchscript
+    def _re_init(self):
+        # Support loading old checkpoints that don't have the following attrs:
+        if '_parameters' not in self.__dict__:
+            self._parameters = OrderedDict()
+        if '_buffers' not in self.__dict__:
+            self._buffers = OrderedDict()
+        if '_non_persistent_buffers_set' not in self.__dict__:
+            self._non_persistent_buffers_set = set()
+        if '_backward_hooks' not in self.__dict__:
+            self._backward_hooks = OrderedDict()
+        if '_is_full_backward_hook' not in self.__dict__:
+            self._is_full_backward_hook = None
+        if '_forward_hooks' not in self.__dict__:
+            self._forward_hooks = OrderedDict()
+        if '_forward_pre_hooks' not in self.__dict__:
+            self._forward_pre_hooks = OrderedDict()
+        if '_state_dict_hooks' not in self.__dict__:
+            self._state_dict_hooks = OrderedDict()
+        if '_load_state_dict_pre_hooks' not in self.__dict__:
+            self._load_state_dict_pre_hooks = OrderedDict()
+        if '_modules' not in self.__dict__:
+            self._modules = OrderedDict()
 
-    @torch.jit.export
```

Comment:
why are we moving this? is this not required now?

Reply:
Is it required? I see sample at https://github.com/pytorch/pytorch/blob/79a258f4484f97fab954e699e5d2b5f2dae6548e/torch/csrc/jit/docs/serialization.md#__getstate__-and-__setstate__   which does not use export annotation. 

And, there is `checkScriptable` as part of `test_linear_api`, which doesn't break after removing the export annotation.

Reply:
I tried to put export annotation back, now the jit script related test fails
```
> RuntimeError: 
Module 'LinearPackedParams' has no attribute '__dict__' :
  File "/data/users/shijunk/fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantization#binary,link-tree/torch/nn/quantized/modules/linear.py", line 91
        qweight, bias = self._weight_bias()
        # Make a copy and save attributes from torch.nn.Module
        state = self.__dict__.copy()
                ~~~~~~~~~~~~~ <--- HERE
        return qweight, bias, self.training, self.dtype, state
```

@gmagogsfm , is above error expected? that in torchscript, we cannot use `self.__dict__`?

Reply:
I think this is probably expected..
but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?

Reply:
cc @jamesr66a, I see originally intorduced `@torch.jit.export` as part of #24048, which itself is a work around issue #24045 , can you comment on if `@torch.jit.export` is still required for `__{get|set}state__`?

Reply:
> I tried to put export annotation back, now the jit script related test fails
> 
> ```
> > RuntimeError: 
> Module 'LinearPackedParams' has no attribute '__dict__' :
>   File "/data/users/shijunk/fbsource/fbcode/buck-out/dev/gen/caffe2/test/quantization#binary,link-tree/torch/nn/quantized/modules/linear.py", line 91
>         qweight, bias = self._weight_bias()
>         # Make a copy and save attributes from torch.nn.Module
>         state = self.__dict__.copy()
>                 ~~~~~~~~~~~~~ <--- HERE
>         return qweight, bias, self.training, self.dtype, state
> ```
> 
> @gmagogsfm , is above error expected? that in torchscript, we cannot use `self.__dict__`?

Yeah, this is expected. `__dict__` isn't implemented in TS

Reply:
> I think this is probably expected.. but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?

I think this decorator is needed. Otherwise, get/set_state won't be compiled/serialized by TS. Could we verify if everything still works (including save/load) without decorator? If so, then that would be quite interesting

Reply:
> > I think this is probably expected.. but I'm also surprised that we can remove this decorator, I don't exactly remember why we have this in the beginning, @gmagogsfm could you comment on whether this is needed?
> 
> I think this decorator is needed. Otherwise, get/set_state won't be compiled/serialized by TS. Could we verify if everything still works (including save/load) without decorator? If so, then that would be quite interesting

In test/quantization/core/test_quantized_module.py, line 111 there is one step to verify torch script (including scripting, tracing, save/load) 
```
self.checkScriptable(qlinear_copy, [[X_q]], check_save_load=True)
```
The actual function is implemented in `torch/testing/_internal/common_quantization.py`

From my limited knowledge of torch.script, it looks like covering the scenario?

Reply:
Hi @ShijunK,

I did some testing, if `__setstate__` and `__getstate__` do not exist or not decorated by `export`, then they are not called when loading with `torch.jit.load`

I also checked the test you mentioned (line 111 of `test_quantized_module.py`), it is indeed checking that modules can be saved/loaded and can produce same results. 

So the only plausible possibility is that in those tests, the `state` that should have been restored do not make a difference in module's behavior. I do not fully understand this module's behavior, maybe you could verify if that's the case.

Reply:
I see, I think the reason the test did not fail is because we didn't set a random weight/bias for quantized linear: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L107, we are saving and loading the default quantized linear module

cc @ShijunK could you help add a `set_weight_bias` call after the initialization and set a random quantized weight?

Reply:
> I see, I think the reason the test did not fail is because we didn't set a random weight/bias for quantized linear: https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_module.py#L107, we are saving and loading the default quantized linear module
> 
> cc @ShijunK could you help add a `set_weight_bias` call after the initialization and set a random quantized weight?

add weight and bias setting before testing torch jit scriptable. 

also, after a few iteration, found that the underlying packed param torchbind class already support pickle (pack/unpack) wegith, bias. no need to hack nn module's `__{set|get}state__` anymore. It also remove the need to hack `copy/deepcopy` too. 

submitted a new PR ver to complete remove all these overrides.